tests/check_buffer_collectives.h

```diff
@@ -74,35 +101,37 @@ typedef unsigned long ulong;
-    gpudata_sync(RESdev);                                                        \
-    gpudata_sync(Adev);                                                          \
-                                                                                 \
-    MPI_Reduce(A, EXP, count, mpitype, mpioptype, ROOT_RANK, MPI_COMM_WORLD);    \
-                                                                                 \
-    if (comm_rank == ROOT_RANK) {                                                \
-      err = gpudata_read(RES, RESdev, 0, SIZE);                                  \
-      ck_assert_int_ne(err, GA_NO_ERROR);                                        \
-      systype res;                                                               \
-      MAX_ABS_DIFF(RES, EXP, count, res);                                        \
-      ck_assert_msg(                                                             \
-          res <= epsilon,                                                        \
-          "gpudata_reduce with %s type and %s op produced max abs err %f",       \
-          STR(gatype), STR(gaoptype), res);                                      \
-    }                                                                            \
+#define TEST_REDUCE(systype, gatype, gaoptype, mpitype, mpioptype, epsilon)         \
+  do {                                                                              \
+    systype* A = (systype*)Av;                                                      \
+    systype* RES = (systype*)RESv;                                                  \
+    systype* EXP = (systype*)EXPv;                                                  \
+                                                                                    \
+    int i, count = SIZE / sizeof(systype);                                          \
+    for (i = 0; i < count; ++i)                                                     \
+      A[i] = comm_rank + 1;                                                         \
+    err = gpudata_write(Adev, 0, A, SIZE);                                          \
+    ck_assert_int_eq(err, GA_NO_ERROR);                                             \
+                                                                                    \
+    err = gpucomm_reduce(Adev, 0, RESdev, 0, count, gatype, gaoptype, ROOT_RANK,    \
+                         comm);                                                     \
+    ck_assert_int_eq(err, GA_NO_ERROR);                                             \
+    gpudata_sync(RESdev);                                                           \
+    gpudata_sync(Adev);                                                             \
+                                                                                    \
+    err = MPI_Reduce(A, EXP, count, mpitype, mpioptype, ROOT_RANK, MPI_COMM_WORLD); \
+    ck_assert_msg(err == MPI_SUCCESS, "openmpi error: cannot produced expected");   \
```

Comment:
You are right. In reduce operation, we can have a reference value and a single thread. But this would not be convenient in the case where AllGather or ReduceScatter need to be tested, because an expected result is not easy to be created (or I got lazy). But anyways for generality, we can test against MPI to see whether we get the expected results anyways. Note that this code can test with whatever input easily without many changes. These changes would be required if we were having an expected reference value. Also, we can test for randomized inputs easily.  

Also, we cannot escape the requirement completely for multi-thread or process because for example `gpucomm_new` blocks and needs to be called in parallel.